### PR TITLE
[Bug] Address go-routines leak

### PR DESCRIPTION
**Pull request by red-0ne**
### Human Summary

`AppGate Server` and `RelayMiner` leak go-routines when handling relays, resulting to an ever increasing number each time a relay is processed.

This PR addresses most of the leaks by cancelling the contexts passed to `ReplayObservables` and `ReplayClients` but does not eliminate them totally.

Means to test and observe these leaks are described in #354.

### AI Summary

reviewpad:summary

## Issue

- #354 

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_develop_and_test`
- [x] **Run E2E tests locally**: `make test_e2e`
- [ ] **Run E2E tests on DevNet**: Add the `devnet-test-e2e` label to the PR. This is VERY expensive, only do it after all the reviews are complete.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: pkg/relayer/session/session.go [hunk 1]
```diff
@@ -225,8 +225,10 @@ func (rp *relayerSessionsManager) validateConfig() error {
 // waitForBlock blocks until the block at the given height (or greater) is
 // observed as having been committed.
 func (rs *relayerSessionsManager) waitForBlock(ctx context.Context, height int64) client.Block {
+	ctx, cancel := context.WithCancel(ctx)
```
Comment (Olshansk) [inline pkg/relayer/session/session.go:228]: #PUC why we need to cancel

File: pkg/relayer/session/session.go [hunk 2]
```diff
@@ -225,8 +225,10 @@ func (rp *relayerSessionsManager) validateConfig() error {
...
+	defer cancel()
+
 	subscription := rs.blockClient.CommittedBlocksSequence(ctx).Subscribe(ctx)
```
Comment (Olshansk) [inline pkg/relayer/session/session.go:231]: Why don't we unsusbscripe anymore? #PUC

File: pkg/observable/channel/replay.go [hunk 3]
```diff
@@ -111,7 +112,18 @@ func (ro *replayObservable[V]) Subscribe(ctx context.Context) observable.Observe
...
-	observer := NewObserver[V](ctx, ro.observerManager.remove)
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	// caller can cancel context or close the publish channel to unsubscribe active observers
+	ctx, cancel := context.WithCancel(ctx)
+	removeAndCancel := func(toRemove observable.Observer[V]) {
+		ro.observerManager.remove(toRemove)
+		cancel()
```
Comment (Olshansk) [inline pkg/observable/channel/replay.go:123]: If the caller never calles `removeAndCancel`, does this never get cancelled?

File: pkg/observable/channel/map.go [hunk 4]
```diff
@@ -112,6 +121,7 @@ func goMapTransformNotification[S, D any](
 	srcObserver observable.Observer[S],
 	transformFn MapFn[S, D],
 	publishFn func(dstNotifications D),
+	doneFn func(),
```
Comment (Olshansk) [inline pkg/observable/channel/map.go:124]: #PUC on what's expected of this function.

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/client/events/replay_client.go [hunk 5]
```diff
@@ -145,21 +145,10 @@ func (rClient *replayClient[T]) EventsSequence(ctx context.Context) observable.R
 // goRemapEventsSequence publishes events observed by the most recent cached
 // events type replay observable to the given publishCh
 func (rClient *replayClient[T]) goRemapEventsSequence(ctx context.Context, publishCh chan<- T) {
-	var prevEventTypeObs observable.ReplayObservable[T]
 	channel.ForEach[observable.ReplayObservable[T]](
 		ctx,
 		rClient.replayObsCache,
 		func(ctx context.Context, eventTypeObs observable.ReplayObservable[T]) {
-			if prevEventTypeObs != nil {
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply pkg/client/events/replay_client.go:153]: This was originally added in reaction to a failing test at some point, IIRC. Looking at it now, I don't think it works like we thought at the time. I think in reality it was a band-aid that was covering up some of the other leaks that this PR fixes which are present deeper in dependency tree (i.e. `Observable` & `ReplayObservable`).

File: pkg/observable/channel/map.go (hunk 4 above)
Comment (bryanchriswhite) [reply pkg/observable/channel/map.go:124]: Could this be a channel instead of an anonymous function? Then we would just call `close` on 134 instead of the func instead.

I think the comment should be something to the effect of: it's the sender's responsibility to close the channel, since the publish channel (`dstProducer`) and its associated observable are constructed in this scope, if they're not also closed here then the associated goroutine will leak.

Comment (bryanchriswhite) [reply pkg/observable/channel/map.go:124]: Could this be a channel instead of an anonymous function? Then we would just call `close` on 134 instead of the func instead.

I think the comment should be something to the effect of: it's the sender's responsibility to close the channel, since the publish channel (`dstProducer`) and its associated observable are constructed in this scope, if they're not also closed here then the associated goroutine will leak. Closing the channel ends the goroutine.

File: pkg/relayer/session/session.go (hunk 2 above)
Comment (bryanchriswhite) [reply pkg/relayer/session/session.go:231]: I think the comment you're asking for already exists in `replayObservable#Subscribe()`:

```
// caller can cancel context or close the publish channel to unsubscribe active observers
```

File: pkg/observable/channel/replay.go (hunk 3 above)
Comment (bryanchriswhite) [reply pkg/observable/channel/replay.go:123]: `removeAndCancel` is only ever called in the `Observer` implementation (see: `#unsubscribe()`, where it's referenced as `observer#onUnsubscribe()`) via [`channelObserverManager#goUnsubscribeOnDone()`](https://github.com/pokt-network/poktroll/blob/ddfdd3ca0331991e98590a1366bcdffc7deee109/pkg/observable/channel/observer_manager.go#L135), below.

File: pkg/relayer/session/session.go (hunk 1 above)
Comment (bryanchriswhite) [reply pkg/relayer/session/session.go:228]: Anywhere we `#Subscribe()` to an observable, we need to either cancel the associated context or close the associated publish channel. The former is preferable for consistency because we may not always have a reference to the publish channel. For example, an observable (and it's corresponding publish channel) could be constructed in another scope that only returns the observable itself. In this case the only way to "close" the observable is to cancel the context.

**Review by bryanchriswhite — APPROVED** [review]
Looks good to me - thanks for diving into this! :raised_hands:

File: pkg/client/events/replay_client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply pkg/client/events/replay_client.go:153]: @bryanchriswhite To your knowledge, do you think we have any more similar bandaids in the code?

File: pkg/relayer/session/session.go (hunk 1 above)
Comment (Olshansk) [reply pkg/relayer/session/session.go:233]: @bryanchriswhite Shold the observable be able to close itself?

File: pkg/client/tx/client.go [hunk 6]
```diff
@@ -375,6 +375,10 @@ func (tClient *txClient) addPendingTransactions(
 // Parameters:
 // - ctx: Context for managing the function's lifecycle and child operations.
 func (tClient *txClient) goSubscribeToOwnTxs(ctx context.Context) {
+	// Create a cancellable child context for managing the EventsSequence lifecycle.
+	// Cancelling it at the end of the function, when listening for submitted
+	// transactions is no longer needed will ensure that the EventsSequence, its
```
Comment (Olshansk) [inline pkg/client/tx/client.go:380]: `is no longer needed will`

Is this missing an `and`?

File: pkg/observable/channel/map.go [hunk 7]
```diff
@@ -28,9 +28,7 @@ func Map[S, D any](
 		func(dstNotification D) {
 			dstProducer <- dstNotification
 		},
-		func() {
-			close(dstProducer)
-		},
+		dstProducer,
```
Comment (Olshansk) [inline pkg/observable/channel/map.go:31]: Nice

File: pkg/observable/channel/map.go [hunk 8]
```diff
@@ -116,12 +110,16 @@ func ForEach[V any](
 
 // goMapTransformNotification transforms, optionally skips, and publishes
 // notifications via the given publishFn.
-func goMapTransformNotification[S, D any](
+func goMapTransformNotification[S, D any, P any](
```
Comment (Olshansk) [inline pkg/observable/channel/map.go:113]: Can we do this?

```suggestion
func goMapTransformNotification[S, D, P any](
```

**Review by Olshansk — APPROVED** [review]
A few small comments but approving regardless.

I can't say I have a 100% understanding of all the observable flows but trust that it's working as expected.

File: pkg/client/tx/client.go (hunk 6 above)
Comment (red-0ne) [reply pkg/client/tx/client.go:380]: It should not. What about:
```
// Cancelling it when listening for submitted transactions is no longer needed
// will ensure that the EventsSequence, its subscriptions and underlying channels
// are effectively closed.
```

Removed the `Cancelling it at the end of the function` since the `defer cancel()` is not far from the comment.

File: pkg/client/tx/client.go [hunk 9]
```diff
@@ -376,9 +376,9 @@ func (tClient *txClient) addPendingTransactions(
 // - ctx: Context for managing the function's lifecycle and child operations.
 func (tClient *txClient) goSubscribeToOwnTxs(ctx context.Context) {
 	// Create a cancellable child context for managing the EventsSequence lifecycle.
-	// Cancelling it at the end of the function, when listening for submitted
-	// transactions is no longer needed will ensure that the EventsSequence, its
-	// subscriptions and underlying channels are effectively closed.
+	// Cancelling it when listening for submitted transactions is no longer needed
```
Comment (Olshansk) [inline pkg/client/tx/client.go:379]: NIT: is/are but we can do it in a separate commit

File: pkg/client/tx/client.go (hunk 6 above)
Comment (Olshansk) [reply pkg/client/tx/client.go]: LGTM
